encode.go

```diff
@@ -11,74 +11,92 @@ import (
 )
 
 // TODO: Handle all the errors
```

Comment:
Maybe a comment here `Negative constants are scalars, positive are vectors`? Then the comments down below could just say the type instead of repeated scalar and vector